### PR TITLE
Initial commit for gardener shoot oidc extension

### LINKED ISSUE #3: Add gardener extension submodule
**What would you like to be added**:

A new Gardener extension. It should have a binary that:

- Should be like any other gardener extension (e.g. https://github.com/gardener/gardener-extension-shoot-dns-service)
- Should 
  - deploy the [oidc-webhook-authenticator](https://github.com/gardener/oidc-webhook-authenticator) in the shoot namespace
  - create needed certificates for authentication
  - mutate kube-apiserver Pods and add the needed flags to them
  - add network policies

### PR DESCRIPTION
**Pull request by dimityrmirchev**
**What this PR does / why we need it**:
This is the initial commit for the Gardener extension shoot oidc service.
**Which issue(s) this PR fixes**:
Fixes #3

**Special notes for your reviewer**:

**Release note**:

```feature user

```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,5 +1,57 @@
...
-TBD
+Project Gardener implements the automated management and operation of [Kubernetes](https://kubernetes.io/) clusters as a service.
+Its main principle is to leverage Kubernetes concepts for all of its tasks.
+
+Recently, most of the vendor specific logic has been developed [in-tree](https://github.com/gardener/gardener).
+However, the project has grown to a size where it is very hard to extend, maintain, and test.
+With [GEP-1](https://github.com/gardener/gardener/blob/master/docs/proposals/01-extensibility.md) we have proposed how the architecture can be changed in a way to support external controllers that contain their very own vendor specifics.
+This way, we can keep Gardener core clean and independent.
+
+This controller implements Gardener's extension contract for the AWS provider.
```
Comment (vpnachev) [inline README.md:13]:
```suggestion
This controller implements Gardener's extension contract for the `extension.extensions.gardener.cloud/v1alpha1` extension.
```

Or 

```suggestion
This controller implements Gardener's extension contract for the shoot-oidc-service Extension.
```

File: README.md [hunk 2]
```diff
@@ -1,5 +1,57 @@
...
+metadata:
+  name: extension-shoot-oidc-service
+  namespace: shoot--project--abc
+spec:
+  type: shoot-oidc-service
+```
+
+When an extension resource is reconciled, the extension controller will create an instance of [OIDC Webhook Authenticator](https://github.com/gardener/oidc-webhook-authenticator). These resources are placed inside the shoot namespace on the seed. Also, the controller takes care about generating necessary `RBAC` resources for the seed as well as for the shoot.
+
+Please note, this extension controller relies on the [Gardener-Resource-Manager](https://github.com/gardener/gardener-resource-manager) to deploy k8s resources to seed and shoot clusters.
```
Comment (vpnachev) [inline README.md:35]:
```suggestion
Please note, this extension controller relies on the [Gardener-Resource-Manager](https://github.com/gardener/gardener/blob/master/docs/concepts/resource-manager.md) to deploy k8s resources to seed and shoot clusters.
```

File: charts/gardener-extension-shoot-oidc-service/templates/service.yaml [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    app.kubernetes.io/instance: {{ .Release.Name }}
+spec:
+  type: ClusterIP
+  selector:
+    app.kubernetes.io/name: gardener-extension-{{ .Values.serviceName }}
+    app.kubernetes.io/instance: {{ .Release.Name }}
+  ports:
+  - port: {{ .Values.webhookConfig.servicePort }}
+    protocol: TCP
+    targetPort: {{ .Values.webhookConfig.serverPort }}
```
Comment (vpnachev) [inline charts/gardener-extension-shoot-oidc-service/templates/service.yaml:26]: Could you clarify what is the difference of service and server port? 
In the deployment I see
```yaml
--webhook-config-service-port={{ .Values.webhookConfig.servicePort }}
--webhook-config-server-port={{ .Values.webhookConfig.serverPort }}
```

Which also confused me, but now I see that actually the one port is mapped to the other via the k8s Service.

Comment (dimityrmirchev) [reply charts/gardener-extension-shoot-oidc-service/templates/service.yaml:26]: The service port is used by the extension framework to register a webhookconfiguration see [here](https://github.com/gardener/gardener/blob/d19723a97065c6f626d1b9b59b458f71d2cce644/extensions/pkg/webhook/cmd/options.go#L244). On the other hand if not specified it is defaulted to the server port see [here](https://github.com/gardener/gardener/blob/d19723a97065c6f626d1b9b59b458f71d2cce644/extensions/pkg/webhook/cmd/options.go#L227). Maybe it makes sense to remove the `--webhook-config-service-port` configuration and leave it to be defaulted by the extension framework. WDYT?

File: example/25-rbac.yaml [hunk 4]
```diff
@@ -0,0 +1,79 @@
+# SPDX-FileCopyrightText: 2021 SAP SE or an SAP affiliate company and Gardener contributors
```
Comment (vpnachev) [inline example/25-rbac.yaml:1]: I am not sure if RBACs are needed here, but
- the permissions here differ from the one in the charts
- there is no binding, so no user is actually using them at the moment.

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+EXTENSION_PREFIX            := gardener-extension
+NAME                        := shoot-oidc-service
+REGISTRY                    := eu.gcr.io/gardener-project/gardener
+IMAGE_PREFIX                := $(REGISTRY)/extensions
+REPO_ROOT                   := $(shell dirname $(realpath $(lastword $(MAKEFILE_LIST))))
+HACK_DIR                    := $(REPO_ROOT)/hack
+VERSION                     := $(shell cat "$(REPO_ROOT)/VERSION")
+LD_FLAGS                    := "-w -X github.com/gardener/oidc-webhook-authenticator/$(EXTENSION_PREFIX)-$(NAME)/pkg/version.Version=$(IMAGE_TAG)"
```
Comment (vpnachev) [inline Makefile:12]:
```suggestion
LD_FLAGS                    := "-w -X github.com/gardener/$(EXTENSION_PREFIX)-$(NAME)/pkg/version.Version=$(IMAGE_TAG)"
```

File: cmd/gardener-extension-shoot-oidc-service/app/app.go [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+	mgr, err := manager.New(o.restOptions.Completed().Config, mgrOpts)
+	if err != nil {
+		return fmt.Errorf("could not instantiate controller-manager: %s", err)
+	}
+
+	if err := extensionscontroller.AddToScheme(mgr.GetScheme()); err != nil {
+		return fmt.Errorf("could not update manager scheme: %s", err)
+	}
+
+	// TODO Should I install v1alphav1 from oidc-webhook-authenticator?
```
Comment (vpnachev) [inline cmd/gardener-extension-shoot-oidc-service/app/app.go:70]: No, this controller is not working with them.

File: pkg/cmd/options.go [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+type OIDCServiceConfig struct {
+	config apisconfig.Configuration
+}
+
+// Apply applies the OIDCServiceOptions to the passed ControllerOptions instance.
+func (c *OIDCServiceConfig) Apply(config *controllerconfig.Config) {
+	config.Configuration = c.config
+}
+
+func (c *OIDCServiceConfig) ApplyHealthCheckConfig(config *healthcheckconfig.HealthCheckConfig) {
```
Comment (vpnachev) [inline pkg/cmd/options.go:101]: Needs doc string.

File: pkg/cmd/options.go [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+	if c.config.HealthCheckConfig != nil {
+		*config = *c.config.HealthCheckConfig
+	}
+}
+
+// ControllerSwitches are the cmd.ControllerSwitches for the provider controllers.
```
Comment (vpnachev) [inline pkg/cmd/options.go:107]:
```suggestion
// ControllerSwitches are the cmd.ControllerSwitches for the extension controllers.
```

File: pkg/controller/lifecycle/actuator.go [hunk 9]
```diff
@@ -0,0 +1,701 @@
...
+// Delete the Extension resource.
+func (a *actuator) Delete(ctx context.Context, ex *extensionsv1alpha1.Extension) error {
+	namespace := ex.GetNamespace()
+
+	if err := managedresources.DeleteForSeed(ctx, a.client, namespace, service.ManagedResourceNamesSeed); err != nil {
+		return err
+	}
+	if err := managedresources.DeleteForShoot(ctx, a.client, namespace, service.ManagedResourceNamesShoot); err != nil {
+		return err
+	}
```
[the remark this answers is not in the document]
Comment (dimityrmirchev) [reply pkg/controller/lifecycle/actuator.go]: Yes, this is probably a better option. I reworked this part of the code.

File: charts/gardener-extension-shoot-oidc-service/templates/service.yaml (hunk 3 above)
Comment (vpnachev) [reply charts/gardener-extension-shoot-oidc-service/templates/service.yaml:26]: Thanks! Let's keep as it is for now.

**Review by vpnachev — APPROVED** [review]
/lgtm

**Pull request merged**